### ISSUE TITLE
multiword_ocr fails to convert to NCNN

### ISSUE THREAD

**Issue by cchadowitz-pf** [jolibrain/deepdetect#1151]
#### Configuration

- Version of DeepDetect:
    - [x] Docker CPU - built from https://github.com/jolibrain/deepdetect/archive/v0.13.0.tar.gz using `docker build --build-arg DEEPDETECT_DEFAULT_MODELS=false -t jolibrain/deepdetect_cpu_pfi:0.13.0 --no-cache -f docker/cpu.Dockerfile .`
- Commit (shown by the server when starting):
This build still doesn't show the commit ref when starting, but I used the release archive listed above:
```
DeepDetect  (dev)
GIT REF: :
COMPILE_FLAGS: USE_CAFFE2=OFF USE_TF=OFF USE_NCNN=ON USE_TORCH=OFF USE_HDF5=ON USE_CAFFE=ON USE_TENSORRT=OFF USE_TENSORRT_OSS=OFF USE_DLIB=OFF USE_CUDA_CV=OFF USE_SIMSEARCH=ON USE_ANNOY=OFF USE_FAISS=ON USE_COMMAND_LINE=ON USE_JSON_API=ON USE_HTTP_SERVER=ON
DEPS_VERSION: OPENCV_VERSION=3.2.0 CUDA_VERSION= CUDNN_VERSION= TENSORRT_VERSION=
```
#### Your question / the problem you're facing:
I'm trying to compare the `word_detect_v2` and `multiword_ocr` Caffe models (as a chain) vs the same models converted to the NCNN backend. The `word_detect_v2` model converted without an error, but the `multiword_ocr` model threw an error and killed the server (see the bottom for the server logs).

#### Error message (if any) / steps to reproduce the problem:

- [x] list of API calls:
For `word_detect_v2`, I used:
```
curl -X PUT "http://localhost:8080/services/word_detect_v2" -d '
{
  "description": "word_detect_v2",
  "mllib": "caffe",
  "type": "supervised",
  "parameters": {
    "input": {
      "connector": "image",
      "width": 512,
      "height": 512
    },
    "mllib": {
      "nclasses": 2,
      "gpu": true
    },
    "output": {}
  },
  "model":{
   "repository":"/opt/models/word_detect_v2/"
  }
}' | jq
```
```
curl -X PUT "http://localhost:8080/services/word_detect_v2_ncnn" -d '
{
  "description": "word_detect_v2",
  "mllib": "ncnn",
  "type": "supervised",
  "parameters": {
    "input": {
      "connector": "image",
      "width": 512,
      "height": 512,
      "mean": [115.44238, 121.15586, 121.03909]
    },
    "mllib": {
      "nclasses": 2,
      "gpu": true
    },
    "output": {}
  },
  "model":{
   "repository":"/opt/models/word_detect_v2/"
  }
}' | jq
```
For `multiword_ocr`, I used:
```
curl -X PUT "http://localhost:8080/services/multiword_ocr" -d '
{
    "description": "multiword_ocr",
    "mllib": "caffe",
    "parameters": {
        "input": {
            "connector": "image",
            "ctc": true,
            "height": 136,
            "width": 220
        },
        "mllib": {
            "nclasses": 69,
            "gpu": true
        }
    },
    "type": "supervised",
    "model":{
     "repository":"/opt/models/multiword_ocr/"
    }
}' | jq
```
and
```
curl -X PUT "http://localhost:8080/services/multiword_ocr_ncnn" -d '
{
    "description": "multiword_ocr",
    "mllib": "ncnn",
    "parameters": {
        "input": {
            "connector": "image",
            "ctc": true,
            "height": 136,
            "width": 220
        },
        "mllib": {
            "nclasses": 69,
            "gpu": true
        }
    },
    "type": "supervised",
    "model":{
     "repository":"/opt/models/multiword_ocr/"
    }
}' | jq
```
- [x] Server log output:

```
[2021-01-22 17:36:29.697] [api] [info] 172.17.0.1 "PUT /services/word_detect_v2_ncnn" 201 1411ms
[2021-01-22 17:36:35.430] [api] [info] could not find  ncnn model, converting caffe model /opt/models/multiword_ocr/model_iter_346000.caffemodelin repository /opt/models/multiword_ocr/
layer ContinuationIndicator not exists or registered
 0# dd::HttpJsonAPI::abort(int) at /opt/deepdetect/src/httpjsonapi.cc:626
 1# 0x00007FEB84B8F040 in /lib/x86_64-linux-gnu/libc.so.6
 2# 0x000055E54846F0C1 in /opt/deepdetect/build/main/dede
 3# 0x000055E548473AB4 in /opt/deepdetect/build/main/dede
 4# 0x000055E5484748DE in /opt/deepdetect/build/main/dede
 5# dd::NCNNLib<dd::ImgNCNNInputFileConn, dd::SupervisedOutput, dd::NCNNModel>::init_mllib(dd::APIData const&) at /opt/deepdetect/src/backends/ncnn/ncnnlib.cc:105
 6# dd::MLService<dd::NCNNLib, dd::ImgNCNNInputFileConn, dd::SupervisedOutput, dd::NCNNModel>::init(dd::APIData const&) at /opt/deepdetect/src/mlservice.h:161
 7# dd::Services::add_service(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, mapbox::util::variant<dd::MLService<dd::CaffeLib, dd::ImgCaffeInputFileConn, dd::SupervisedOutput, dd::CaffeModel>, dd::MLService<dd::CaffeLib, dd::CSVCaffeInputFileConn, dd::SupervisedOutput, dd::CaffeModel>, dd::MLService<dd::CaffeLib, dd::CSVTSCaffeInputFileConn, dd::SupervisedOutput, dd::CaffeModel>, dd::MLService<dd::CaffeLib, dd::TxtCaffeInputFileConn, dd::SupervisedOutput, dd::CaffeModel>, dd::MLService<dd::CaffeLib, dd::SVMCaffeInputFileConn, dd::SupervisedOutput, dd::CaffeModel>, dd::MLService<dd::CaffeLib, dd::ImgCaffeInputFileConn, dd::UnsupervisedOutput, dd::CaffeModel>, dd::MLService<dd::CaffeLib, dd::CSVCaffeInputFileConn, dd::UnsupervisedOutput, dd::CaffeModel>, dd::MLService<dd::CaffeLib, dd::CSVTSCaffeInputFileConn, dd::UnsupervisedOutput, dd::CaffeModel>, dd::MLService<dd::CaffeLib, dd::TxtCaffeInputFileConn, dd::UnsupervisedOutput, dd::CaffeModel>, dd::MLService<dd::CaffeLib, dd::SVMCaffeInputFileConn, dd::UnsupervisedOutput, dd::CaffeModel>, dd::MLService<dd::XGBLib, dd::CSVXGBInputFileConn, dd::SupervisedOutput, dd::XGBModel>, dd::MLService<dd::XGBLib, dd::SVMXGBInputFileConn, dd::SupervisedOutput, dd::XGBModel>, dd::MLService<dd::XGBLib, dd::TxtXGBInputFileConn, dd::SupervisedOutput, dd::XGBModel>, dd::MLService<dd::TSNELib, dd::CSVTSNEInputFileConn, dd::UnsupervisedOutput, dd::TSNEModel>, dd::MLService<dd::TSNELib, dd::TxtTSNEInputFileConn, dd::UnsupervisedOutput, dd::TSNEModel>, dd::MLService<dd::NCNNLib, dd::CSVTSNCNNInputFileConn, dd::SupervisedOutput, dd::NCNNModel>, dd::MLService<dd::NCNNLib, dd::ImgNCNNInputFileConn, dd::SupervisedOutput, dd::NCNNModel> >&&, dd::APIData const&) at /opt/deepdetect/src/services.h:399
 8# dd::JsonAPI::service_create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) at /opt/deepdetect/src/jsonapi.cc:685
 9# APIHandler::operator()(boost::network::http::basic_request<boost::network::http::tags::http_server> const&, boost::network::http::basic_response<boost::network::http::tags::http_server>&) at /opt/deepdetect/src/httpjsonapi.cc:444
10# boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>::handle_read_headers(boost::system::error_code const&, unsigned long) at /opt/deepdetect/build/cpp-netlib/src/cpp-netlib/boost/network/protocol/http/server/sync_connection.hpp:139
11# void boost::asio::detail::strand_service::dispatch<boost::asio::detail::binder2<boost::_bi::bind_t<void, boost::_mfi::mf2<void, boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>, boost::system::error_code const&, unsigned long>, boost::_bi::list3<boost::_bi::value<boost::shared_ptr<boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler> > >, boost::arg<1> (*)(), boost::arg<2> (*)()> >, boost::system::error_code, unsigned long> >(boost::asio::detail::strand_service::strand_impl*&, boost::asio::detail::binder2<boost::_bi::bind_t<void, boost::_mfi::mf2<void, boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>, boost::system::error_code const&, unsigned long>, boost::_bi::list3<boost::_bi::value<boost::shared_ptr<boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler> > >, boost::arg<1> (*)(), boost::arg<2> (*)()> >, boost::system::error_code, unsigned long>&) at /usr/include/boost/asio/detail/impl/strand_service.hpp:91
12# void boost::asio::detail::wrapped_handler<boost::asio::io_service::strand, boost::_bi::bind_t<void, boost::_mfi::mf2<void, boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>, boost::system::error_code const&, unsigned long>, boost::_bi::list3<boost::_bi::value<boost::shared_ptr<boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler> > >, boost::arg<1> (*)(), boost::arg<2> (*)()> >, boost::asio::detail::is_continuation_if_running>::operator()<boost::system::error_code, unsigned long>(boost::system::error_code const&, unsigned long const&) at /usr/include/boost/asio/detail/wrapped_handler.hpp:99
13# boost::asio::detail::completion_handler<boost::asio::detail::rewrapped_handler<boost::asio::detail::binder2<boost::asio::detail::wrapped_handler<boost::asio::io_service::strand, boost::_bi::bind_t<void, boost::_mfi::mf2<void, boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>, boost::system::error_code const&, unsigned long>, boost::_bi::list3<boost::_bi::value<boost::shared_ptr<boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler> > >, boost::arg<1> (*)(), boost::arg<2> (*)()> >, boost::asio::detail::is_continuation_if_running>, boost::system::error_code, unsigned long>, boost::_bi::bind_t<void, boost::_mfi::mf2<void, boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>, boost::system::error_code const&, unsigned long>, boost::_bi::list3<boost::_bi::value<boost::shared_ptr<boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler> > >, boost::arg<1> (*)(), boost::arg<2> (*)()> > > >::do_complete(boost::asio::detail::task_io_service*, boost::asio::detail::task_io_service_operation*, boost::system::error_code const&, unsigned long) at /usr/include/boost/asio/detail/completion_handler.hpp:59
14# void boost::asio::detail::strand_service::dispatch<boost::asio::detail::rewrapped_handler<boost::asio::detail::binder2<boost::asio::detail::wrapped_handler<boost::asio::io_service::strand, boost::_bi::bind_t<void, boost::_mfi::mf2<void, boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>, boost::system::error_code const&, unsigned long>, boost::_bi::list3<boost::_bi::value<boost::shared_ptr<boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler> > >, boost::arg<1> (*)(), boost::arg<2> (*)()> >, boost::asio::detail::is_continuation_if_running>, boost::system::error_code, unsigned long>, boost::_bi::bind_t<void, boost::_mfi::mf2<void, boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>, boost::system::error_code const&, unsigned long>, boost::_bi::list3<boost::_bi::value<boost::shared_ptr<boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler> > >, boost::arg<1> (*)(), boost::arg<2> (*)()> > > >(boost::asio::detail::strand_service::strand_impl*&, boost::asio::detail::rewrapped_handler<boost::asio::detail::binder2<boost::asio::detail::wrapped_handler<boost::asio::io_service::strand, boost::_bi::bind_t<void, boost::_mfi::mf2<void, boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>, boost::system::error_code const&, unsigned long>, boost::_bi::list3<boost::_bi::value<boost::shared_ptr<boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler> > >, boost::arg<1> (*)(), boost::arg<2> (*)()> >, boost::asio::detail::is_continuation_if_running>, boost::system::error_code, unsigned long>, boost::_bi::bind_t<void, boost::_mfi::mf2<void, boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>, boost::system::error_code const&, unsigned long>, boost::_bi::list3<boost::_bi::value<boost::shared_ptr<boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler> > >, boost::arg<1> (*)(), boost::arg<2> (*)()> > >&) at /usr/include/boost/asio/detail/impl/strand_service.hpp:85
15# boost::asio::detail::reactive_socket_recv_op<boost::asio::mutable_buffers_1, boost::asio::detail::wrapped_handler<boost::asio::io_service::strand, boost::_bi::bind_t<void, boost::_mfi::mf2<void, boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler>, boost::system::error_code const&, unsigned long>, boost::_bi::list3<boost::_bi::value<boost::shared_ptr<boost::network::http::sync_connection<boost::network::http::tags::http_server, APIHandler> > >, boost::arg<1> (*)(), boost::arg<2> (*)()> >, boost::asio::detail::is_continuation_if_running> >::do_complete(boost::asio::detail::task_io_service*, boost::asio::detail::task_io_service_operation*, boost::system::error_code const&, unsigned long) at /usr/include/boost/asio/detail/reactive_socket_recv_op.hpp:110
16# boost::network::http::sync_server_base<boost::network::http::tags::http_server, APIHandler>::run() at /opt/deepdetect/build/cpp-netlib/src/cpp-netlib/boost/network/protocol/http/server/sync_server.hpp:49
17# 0x00007FEB857E36DF in /usr/lib/x86_64-linux-gnu/libstdc++.so.6
18# 0x00007FEB891056DB in /lib/x86_64-linux-gnu/libpthread.so.0
19# 0x00007FEB84C7171F in /lib/x86_64-linux-gnu/libc.so.6
```

I can attach the full server log if necessary, but it's mostly just the Caffe output.

**Comment by beniz** [comment]
Ah bummer this is a custom layer we have. It's possibly a bug of our patches for ncnn. It'll take another round to fix, this means our OCR models are broken with ncnn at the moment.

**Comment by cchadowitz-pf** [comment]
👍 glad you know where it's coming from! :)

**Comment by beniz** [comment]
OK, so this first comes from a missing patch, and #1152 remediates to the missing patch.
Now this appears to solve the OCR model conversion, but inference yields another crash: 

```
[2021-01-22 20:54:54.088] [test] [info] [0] / executing predict on service word_detect
[New Thread 0x7fffab674700 (LWP 4780)]
[New Thread 0x7fffaae73700 (LWP 4781)]
[New Thread 0x7fffaa672700 (LWP 4782)]
[New Thread 0x7fff99673700 (LWP 4783)]
[New Thread 0x7fff98e72700 (LWP 4784)]
[New Thread 0x7fff8dffd700 (LWP 4785)]
[New Thread 0x7fff8d7fc700 (LWP 4786)]
[New Thread 0x7fffb2e7b700 (LWP 4787)]
[New Thread 0x7fffafc78700 (LWP 4788)]
[New Thread 0x7fffa9e71700 (LWP 4789)]
[New Thread 0x7fff98671700 (LWP 4790)]
[New Thread 0x7fff8cffb700 (LWP 4791)]
[New Thread 0x7fff8cbfa700 (LWP 4792)]
[New Thread 0x7fff8c7f9700 (LWP 4793)]
[2021-01-22 20:55:20.418] [test] [info] [1] / executing action crop
[2021-01-22 20:55:20.420] [test] [info] [2] / executing predict on service word_ocr

Thread 38 "dede" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff8dffd700 (LWP 4785)]
0x00007ffff4dcef3a in std::__detail::_List_node_base::_M_unhook() ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007ffff4dcef3a in std::__detail::_List_node_base::_M_unhook() ()
   from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x000055555585a211 in ncnn::UnlockedPoolAllocator::fastMalloc(unsigned long) ()
#2  0x0000555555950012 in ncnn::LSTM::forward(ncnn::Mat const&, ncnn::Mat const&, ncnn::Mat&, ncnn::Option const&, bool) const ()
#3  0x00005555559515a9 in ncnn::LSTM::forward(std::vector<ncnn::Mat, std::allocator<ncnn::Mat> > const&, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const ()
#4  0x0000555555959a87 in ncnn::LSTM_x86_avx2::forward(std::vector<ncnn::Mat, std::allocator<ncnn::Mat> > const&, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const ()
#5  0x000055555586e85a in ncnn::Net::forward_layer(int, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const ()
#6  0x000055555586eb98 in ncnn::Net::forward_layer(int, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const ()
#7  0x000055555586f211 in ncnn::Net::forward_layer(int, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const ()
#8  0x000055555586f211 in ncnn::Net::forward_layer(int, std::vector<ncnn::Mat, std::allocator<ncnn::Mat> >&, ncnn::Option const&) const ()
#9  0x0000555555870856 in ncnn::Extractor::extract(int, ncnn::Mat&, int) ()
#10 0x000055555567d831 in dd::NCNNLib<dd::ImgNCNNInputFileConn, dd::SupervisedOutput, dd::NCNNModel>::predict
    (this=<optimized out>, ad=..., out=...)
    at /home/beniz/projects/deepdetect/dev/deepdetect/src/backends/ncnn/ncnnlib.cc:282
#11 0x00007ffff454a96e in ?? () from /usr/lib/x86_64-linux-gnu/libgomp.so.1
#12 0x00007ffff6a926db in start_thread (arg=0x7fff8dffd700) at pthread_create.c:463
#13 0x00007ffff404c71f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

The crash comes from within ncnn, so more debugging is needed. cc @fantes

**Comment by cchadowitz-pf** [comment]
Is that new crash during inference related to NCNN upstream changes? I hadn't run into that in the past with any NCNN tests I ran but maybe there's more involved than I'm aware of...

**Comment by beniz** [comment]
Maybe, we had to refactor our LSTM patches to NCNN. CI has tests on NCNN LSTMs for time-series, not OCR. So it's possible something was or is newly wrong. We'll look at it in coming days.

**Comment by fantes** [comment]
Hi there

here is another fix : https://github.com/jolibrain/deepdetect/pull/1153
i cannot reproduce segfault from @beniz  , but with this patch ocr works here 
this PR contains continuation indicator layer + fix described below
FTR : ncnn now uses a lot of architecture dependent optimizations , and our patch on innerproducts for keeping h dimension  is implemented only for generic code thus we force generic code execution ; before this patch, data was packed for some architecture optim  but generic code was called. This fix by disabling data packing

**Comment by cchadowitz-pf** [comment]
Awesome! To be clear, was NCNN in DeepDetect *ever* using architecture dependent optimizations, or always using generic code execution?

**Comment by fantes** [comment]
Before merge of https://github.com/jolibrain/deepdetect/pull/1103 we used our old fork https://github.com/jolibrain/ncnn,  that has not been updated since a while. There are a few architecture dependent snippets in it (see subfolders of src/layer https://github.com/jolibrain/ncnn/tree/master/src/layer).
Since, there have been a lot of architecture dependent code added, see inflation of subfolders of src/layer https://github.com/tencent/ncnn/tree/master/src/layer  and large additions of  of #ifdef for different flags (AVX,SSE2 ... )
So, the answer is ... yes and no :) For vanilla ncnn layers, optimizations are used, and much more are used since we use a more recent version, for lstm and innerproduct we implemented some changes to support ocr, lstm , but implemented them only for generic implementation (and so bypass architecture dependent optimizations). 
Finally, we do nothing special for using optimized version, they are selected automatically at compile time by ncnn cmakelists.
